src/geopackage/wkb.cpp

```diff
@@ -0,0 +1,306 @@
+#include "WKB.hpp"
+
+using namespace geopackage;
+
+// ----------------------------------------------------------------------------
+// WKB Readers
+// ----------------------------------------------------------------------------
+
+typename wkb::point_t wkb::read_point(const byte_vector& buffer, int& index, uint8_t order)
+{
+    double x, y;
+    utils::copy_from(buffer, index, x, order);
+    utils::copy_from(buffer, index, y, order);
+    return point_t{x, y};
+}
+
+// ----------------------------------------------------------------------------
+
+typename wkb::linestring_t wkb::read_linestring(const byte_vector& buffer, int& index, uint8_t order)
+{
+    uint32_t count;
+    utils::copy_from(buffer, index, count, order);
+
+    linestring_t linestring;
+    linestring.resize(count);
+    for (auto& child : linestring) {
+        child = read_point(buffer, index, order);
+    }
+
+    return linestring;
+}
+
+// ----------------------------------------------------------------------------
+
+typename wkb::polygon_t wkb::read_polygon(const byte_vector& buffer, int& index, uint8_t order)
+{
+    uint32_t count;
+    utils::copy_from(buffer, index, count, order);
+
+    polygon_t polygon;
+    
+    if (count > 1) {
+        // polygons only have 1 outer ring,
+        // so any extra vectors are considered to be
+        // inner rings.
+        polygon.inners().resize(count - 1);
+    }
+
+    auto outer = read_linestring(buffer, index, order);
+    polygon.outer().reserve(outer.size());
+    for (auto& p : outer) {
+        polygon.outer().push_back(p);
+    }
+
+    for (uint32_t i = 0; i < count - 1; i++) {
+        auto inner = read_linestring(buffer, index, order);
+        polygon.inners().at(i).reserve(inner.size());
+        for (auto& p : inner) {
+            polygon.inners().at(i).push_back(p);
+        }
+    }
+
+    return polygon;
+}
+
+// ----------------------------------------------------------------------------
+
+typename wkb::multipoint_t wkb::read_multipoint(const byte_vector& buffer, int& index, uint8_t order)
+{
+    uint32_t count;
+    utils::copy_from(buffer, index, count, order);
+
+    multipoint_t mp;
+    mp.resize(count);
+
+    for (auto& point : mp) {
+        const byte_t new_order = buffer[index];
+        index++;
+
+        uint32_t type;
```

Comment:
Do I understand correctly that `type` is being read out of the buffer and ignored? If so, do we need some sort of guard that its value matches expectations?

Reply:
Ah, good catch. That's another remnant of older code. We should have a guard there to check that `type` is as expected.

Reply:
Implemented guards in d0bf396.